cranelift-interpreter/src/interpreter.rs

```diff
@@ -0,0 +1,295 @@
+        &self,
+        func_ref: FuncRef,
+        arguments: &[Value],
+    ) -> Result<ControlFlow, Trap> {
+        match self.env.get_by_func_ref(func_ref) {
+            None => Err(Trap::InvalidFunctionReference(func_ref)),
+            Some(func) => self.call(func, arguments),
+        }
+    }
+
+    fn call(&self, function: &Function, arguments: &[Value]) -> Result<ControlFlow, Trap> {
+        debug!("Call: {}({:?})", function.name, arguments);
+        let first_ebb = function.layout.ebbs().next().expect("to have a first ebb");
+        let parameters = function.dfg.ebb_params(first_ebb);
+        let mut frame = Frame::new(function).with_parameters(parameters, arguments);
+        let result = self.block(&mut frame, first_ebb);
+        result
+    }
+
+    fn block(&self, frame: &mut Frame, ebb: Ebb) -> Result<ControlFlow, Trap> {
+        debug!("Block: {}", ebb);
+        for inst in frame.function.layout.ebb_insts(ebb) {
+            match self.inst(frame, inst)? {
+                ControlFlow::Continue => continue,
+                ControlFlow::ContinueAt(ebb, old_names) => {
+                    let new_names = frame.function.dfg.ebb_params(ebb);
+                    frame.rename(&old_names, new_names);
+                    return self.block(frame, ebb); // TODO check that TCO happens
+                }
+                ControlFlow::Return(rs) => return Ok(ControlFlow::Return(rs)),
+            }
+        }
+        Err(Trap::Unreachable)
+    }
+
+    fn binary(
+        &self,
+        frame: &mut Frame,
+        op: fn(Value, Value) -> Value,
+        a: ValueRef,
+        b: ValueRef,
+        r: ValueRef,
+    ) {
+        let a = frame.get(&a);
+        let b = frame.get(&b);
+        let c = op(a.clone(), b.clone());
+        frame.set(r, c);
+    }
+
+    // TODO refactor to only one `binary` method
+    fn binary_imm(
+        &self,
+        frame: &mut Frame,
+        op: fn(Value, Value) -> Value,
+        a: ValueRef,
+        b: Value,
+        r: ValueRef,
+    ) {
+        let a = frame.get(&a);
+        let c = op(a.clone(), b);
+        frame.set(r, c);
+    }
+
+    fn iconst(&self, frame: &mut Frame, imm: Imm64, r: ValueRef) {
+        frame.set(r, Value::Int(imm.into()));
+    }
+
+    fn bconst(&self, frame: &mut Frame, imm: bool, r: ValueRef) {
+        frame.set(r, Value::Bool(imm));
+    }
+
+    // TODO add load/store
+    fn inst(&self, frame: &mut Frame, inst: Inst) -> Result<ControlFlow, Trap> {
+        use ControlFlow::{Continue, ContinueAt};
+        debug!("Inst: {}", &frame.function.dfg.display_inst(inst, None));
+
+        let data = &frame.function.dfg[inst];
+        match data {
+            Binary { opcode, args } => match opcode {
+                Iadd => {
+                    // TODO trap if arguments are of the wrong type; here and below
+                    let res = first_result(frame.function, inst);
+                    self.binary(frame, Add::add, args[0], args[1], res);
```

Comment:
Alternatively, could the `opcode` be passed to `binary`, and the result can be computed there (with a match on the opcode), without passing a function pointer? Then we don't need the implicit add/sub impl of values, and it makes it easier to add new opcodes.

Reply:
In the rebased commits I used a macro instead; let me know what you think of that.